handlers/other/ec2-cleanup.rb

```diff
@@ -79,64 +148,105 @@
 # LICENSE for details
 
 require 'timeout'
-require 'rubygems' if RUBY_VERSION < '1.9.0'
 require 'sensu-handler'
-require 'aws-sdk-core'
-require 'json'
-require 'open-uri'
+require 'net/http'
+require 'uri'
+require 'aws-sdk'
+#require 'sensu-plugins-aws'
 
 class Ec2Node < Sensu::Handler
+  #include Common
+
   def filter; end
 
+  # Method handle
   def handle
-    # #YELLOW
-    unless ec2_node_exists? # rubocop:disable UnlessElse
+    # Call ec2_node_should_be_deleted method and check for instance state and if valid delete from the sensu API otherwise
+    # instance is in invalid state
+    if ec2_node_should_be_deleted?
       delete_sensu_client!
     else
-      puts "[EC2 Node] #{@event['client']['name']} exists in EC2"
+      puts "[EC2 Node] #{instance_id} is in an invalid state"
     end
   end
 
+  # Method to delete client from sensu API
   def delete_sensu_client!
     response = api_request(:DELETE, '/clients/' + @event['client']['name']).code
     deletion_status(response)
   end
 
-  def ec2_node_exists?
-    instance_ids = ec2.describe_instances(filters: [ { name: "instance-state-name", values: acquire_valid_states } ]).reservations.collect { |r| r.instances.map(&:instance_id) }.flatten
-    # Strip the service name off so we can find it in EC2
-    @event['client']['ec2-name'] = @event['client']['name'].scan(/-(i-\w+)$/)[0][0]
-    instance_ids.each do |instance_id|
-      return true if instance_id == @event['client']['ec2-name']
+  def instance_id
+    return @instance_id if @instance_id
+
+    scan_result = @event['client']['name'].scan(/-(i-\w+)$/)
+    @instance_id =
+      if scan_result[0]
+        scan_result[0][0]
+      else
+        @event['client']['name']
+      end
+    @instance_id
+  end
+
+  # Method to check if there is any insance and if instance is in a valid state that could be deleted
+  def ec2_node_should_be_deleted?
+    # Defining region for aws SDK object
+    ec2 = Aws::EC2::Client.new(region: region)
+    settings['ec2_node'] = {} unless settings['ec2_node']
+    instance_states = @event['client']['ec2_states'] || settings['ec2_node']['ec2_states'] || ['shutting-down', 'terminated', 'stopping', 'stopped']
+    instance_reasons = @event['client']['ec2_state_reasons'] || settings['ec2_node']['ec2_state_reasons'] || %w(Client.UserInitiatedShutdown Server.SpotInstanceTermination Client.InstanceInitiatedShutdown)
+
+    begin
+      # Finding the instance
+      instances = ec2.describe_instances(instance_ids: [instance_id]).reservations[0]
```

Comment:
We might need to `flatten` all the reservations, rather than just taking the first...